connectors/sources/salesforce.py

```diff
@@ -0,0 +1,1475 @@
+                )
+                record["ConvertedContact"] = sobjects_by_id["Contact"].get(
+                    record.get("ConvertedContactId"), {}
+                )
+                record["ConvertedOpportunity"] = sobjects_by_id["Opportunity"].get(
+                    record.get("ConvertedOpportunityId"), {}
+                )
+                yield self.doc_mapper.map_lead(record), content_documents
+
+    async def get_campaigns(self):
+        if not await self._is_queryable("Campaign"):
+            self._logger.info(
+                "Object Campaign is not queryable, so they won't be ingested."
+            )
+            return
+
+        query = await self._campaigns_query()
+        async for records in self._yield_non_bulk_query_pages(query):
+            for record in records:
+                content_documents = self._parse_content_documents(record)
+                yield self.doc_mapper.map_campaign(record), content_documents
+
+    async def get_cases(self):
+        if not await self._is_queryable("Case"):
+            self._logger.info(
+                "Object Case is not queryable, so they won't be ingested."
+            )
+            return
+
+        query = await self._cases_query()
+        async for records in self._yield_non_bulk_query_pages(query):
+            case_feeds_by_case_id = {}
+            if await self._is_queryable("CaseFeed") and records:
+                case_ids = [x.get("Id") for x in records]
+                case_feeds = await self.get_case_feeds(case_ids)
+
+                # groupby requires pre-sorting apparently
+                case_feeds.sort(key=lambda x: x["ParentId"])
+                case_feeds_by_case_id = {
+                    k: list(feeds)
+                    for k, feeds in groupby(case_feeds, key=lambda x: x["ParentId"])
+                }
+
+            for record in records:
+                content_documents = self._parse_content_documents(record)
+
+                record["Feeds"] = case_feeds_by_case_id.get(record.get("Id"))
+                yield self.doc_mapper.map_case(record), content_documents
+
+    async def get_case_feeds(self, case_ids):
+        query = await self._case_feeds_query(case_ids)
+        return await self._execute_non_paginated_query(query)
+
+    async def download_content_documents(self, content_documents):
+        for content_document in content_documents:
+            content_version = content_document.get("LatestPublishedVersion", {}) or {}
+            doc = self.doc_mapper.map_content_document(
+                content_document, content_version
+            )
+
+            download_url = content_version.get("VersionDataUrl")
+            if content_version.get("VersionDataUrl"):
+                doc["_attachment"] = await self._download(download_url)
+
+            yield doc
+
+    async def queryable_sobjects(self):
+        """Cached async property"""
+        if self._queryable_sobjects is not None:
```

Comment:
This can be simplified by `@cached_property`

Reply:
It's asynchronous though. My understanding is that cached_property doesn't support async calls.
https://asyncstdlib.readthedocs.io/en/stable/source/api/functools.html#async-caches

Reply:
TIL. Didn't realize that.

Reply:
Sorry I deleted my comment because I think I'm wrong? cached_property doesn't work from functools but the very library I linked has an async cached_property
(linking again) https://asyncstdlib.readthedocs.io/en/stable/source/api/functools.html#async-caches
We don't use this library though so I don't think it's worth introducing. So it IS possible to cache, just not with functools.